README.md
# Biostatistics

LearnR tutorials for biostatistics. 
    
Subjects range from basic descriptive statistics through to complex linear modelling. The tutorials include text, videos, interactive coding exercises and multiple choice quizzes.

These tutorials are suitable for statistical beginners as well as more advanced students. Some basic R coding knowledge is assumed but advanced knowledge is not necessary.

## Using the tutorials
The easiest way to run these tutorials is to make sure you have an up-to-date release of RStudio. Install the package and restart and the tutorials should appear in the Tutorial tab in the pane with Environment, History etc.

If you would rather not use RStudio then you can use the `learnr::runtutorial()` function which will open the tutorial in a separate browser window.

## Datasets
The package includes 19 datasets which are used in the tutorials. These data are mostly publicly available data archived from published research papers.

## Author and licensing

This package is authored by Rob Knell, SBCS, Queen Mary Univerisity of London. Email r.knell@qmul.ac.uk.

The content of the package is released under a GPL-3 license.

## List of data sources

### Frequency Histograms:

Healy, K., *et al.* (2014) Ecology and mode-of-life explain lifespan variation in birds and mammals. Proceedings. Biological sciences, 281, 20140298

### Descriptive Statistics:

Pawłowski, B., Nowak, J., Borkowska, B., Augustyniak, D. & Drulis-Kawa, Z. (2017) Body height and immune efficacy: testing body stature as a signal of biological quality. Proceedings. Biological sciences / The Royal Society, 284: 0171372

### Scatterplots

Charters, J.E., Heiniger, J., Clemente, C.J., Cameron, S.F., Amir Abdul Nasir, A.F., Niehaus, A.C. & Wilson, R.S. (2018) Multidimensional analyses of physical performance reveal a size‐dependent trade‐off between suites of traits. Functional ecology, 32, 1541–1553.

Rylance, J., Chimpini, C., Semple, S., Russell, D.G., Jackson, M.J., Heyderman, R.S. & Gordon, S.B. (2015) Chronic Household Air Pollution Exposure Is Associated with Impaired Alveolar Macrophage Function in Malawian Non-Smokers. PloS one, 10, e0138762.

Burger, J.R., George, M.A., Leadbetter, C. & Shaikh, F. (2019) The allometry of brain size in mammals. Journal of mammalogy, 100, 276–283. 

### Paired t-tests

Andriolli, F.S., Ishikawa, N.K., Vargas-Isla, R., Cabral, T.S., de Bekker, C. & Baccaro, F.B. (2019) Do zombie ant fungi turn their hosts into light seekers? Behavioral ecology 30, 609--616.

### Two-sample t-tests

Cram, D.L., Blount, J.D. & Young, A.J. (2015) The oxidative costs of reproduction are group-size dependent in a wild cooperative breeder. Proceedings of the Royal Society B: Biological sciences, 282: 20152031.

### Chi-square

Sun, Z., Xu, L., Schmid, B.V., Dean, K.R., Zhang, Z., Xie, Y., Fang, X., Wang, S., Liu, Q., Lyu, B., Wan, X., Xu, J., Stenseth, N.C. & Xu, B. (2019) Human plague system associated with rodent diversity and other environmental factors. Royal Society open science, 6, 190216.

Majoris, J.E., D'Aloia, C.C., Francis, R.K. & Buston, P.M. (2018) Differential persistence favors habitat preferences that determine the distribution of a reef fish. Behavioral ecology: official journal of the International Society for Behavioral Ecology, 29, 429--439.

### Correlation

Manczinger, M., Boross, G., Kemény, L., Müller, V., Lenz, T.L., Papp, B. & Pál, C. (2019) Pathogen diversity drives the evolution of generalist MHC-II alleles in human populations. *PLoS biology*, **17**, e3000131.

### Linear models 1 ANOVA

Wang, M., Ruan, W., Kostenko, O., Carvalho, S., Hannula, S.E., Mulder, P.P.J., Bu, F., van der Putten, W.H. & Bezemer, T.M. (2019) Removal of soil biota alters soil feedback effects on plant growth and defense chemistry. The New phytologist, 221, 1478–1491.

### Linear models 2 Regression

LaManna, J.A. *et al.* (2017) Plant diversity increases with the strength of negative density dependence at the global scale. Science, 356, 1389–1392.

Koerner, S.E., Poulsen, J.R., Blanchard, E.J., Okouyi, J. & Clark, C.J. (2017) Vertebrate community composition and diversity declines along a defaunation gradient radiating from rural villages in Gabon. The Journal of applied ecology, 54, 805–814.

### Linear models 4 multi-factor ANOVA

McCarthy, D.M., Morgan, T.J., Jr, Lowe, S.E., Williamson, M.J., Spencer, T.J., Biederman, J. & Bhide, P.G. (2018) Nicotine exposure of male mice produces behavioral impairment in multiple generations of descendants. PLoS biology, 16, e2006497.

Robinson, C.D., Klein, H.S., Murphy, K.D., Parthasarathy, R., Guillemin, K. & Bohannan, B.J.M. (2018) Experimental bacterial adaptation to the zebrafish gut reveals a primary role for immigration. PLoS biology, 16, e2006893.

### Linear models 5 Multiple Regression

Hill, G.E., Hood, W.R., Ge, Z., Grinter, R., Greening, C., Johnson, J.D., Park, N.R., Taylor, H.A., Andreasen, V.A., Powers, M.J., Justyn, N.M., Parry, H.A., Kavazis, A.N. & Zhang, Y. (2019) Plumage redness signals mitochondrial function in the house finch. Proceedings of the Royal Sociaty B: Biological sciences, 286, 20191354.

Houslay, T.M., Houslay, K.F., Rapkin, J., Hunt, J. & Bussière, L.F. (2017) Mating opportunities and energetic constraints drive variation in age‐dependent sexual signalling (ed C Miller). Functional ecology, 31, 728–741.

### Linear models 6 Factors & Continuous Variables

Okada, Y., Katsuki, M., Okamoto, N., Fujioka, H. & Okada, K. (2019) A specific type of insulin-like peptide regulates the conditional growth of a beetle weapon. PLoS biology, 17, e3000541.

Fitzpatrick, J.L., Almbro, M., Gonzalez-Voyer, A., Hamada, S., Pennington, C., Scanlan, J. & Kolm, N. (2012) Sexual selection uncouples the evolution of brain and body size in pinnipeds. Journal of evolutionary biology, 25, 1321--1330.

NEWS.md
# Biostatistics News

1.0.1 Fixed error in confidence intervals tutorial, expanded Introduction.

1.0.0 Package ready for release

inst/tutorials/10 Two sample t-tests/script_for_2_sample_t_test_video.R
set.seed(12)

Control <- round(rnorm(10, mean = 13, sd = 2), 1)
Treatment <- round(rnorm(10, mean = 15.7, sd = 2), 1)

Treatment[4] <- 10.2

Group <- rep(c("Control", "Treatment"), each = 10)
Leaf_mass <- c(Control, Treatment)

boxplot(Leaf_mass ~ Group, xlab = "", ylab = "Leaf mass (g)",
        col = "aquamarine4")


library(gplots)
plotmeans(Leaf_mass ~ Group,
          n.label = FALSE,
          connect = FALSE,
          pch = 16,
          cex = 1.8,
          xlab = "",
          ylab = "Leaf mass (g)",
          col = "aquamarine4",
          barcol = "aquamarine4")

2*(1-pt(2.84, df = 18))


t.test(Leaf_mass ~ Group, var.equal = TRUE)




X1 <- seq(-4, 4, length = 200)
Y1 <- dt(X1, 18)
plot(
  X1,
  Y1,
  type = "n",
  xlab = "x",
  ylab = "P(x)",
  ylim = c(0, 0.42),
  main = "t-distribution on 18 df"
)

x0 <- min(which(X1 >= qt(0.005, 18)))
x1 <- min(which(X1 >= qt(0.025, 18)))
x2 <- min(which(X1 >= qt(0.1666, 18)))
x3 <- max(which(X1 <= qt(0.8333, 18)))
x4 <- max(which(X1 <= qt(0.975, 18)))
x5 <- max(which(X1 <= qt(0.995, 18)))

polygon(
  x = c(X1[c(1, 1:x0, x0)]),
  y = c(0, Y1[1:x0], 0),
  col = "#3182bd",
  border = NA
)
polygon(
  x = c(X1[c(x0, x0:x1, x1)]),
  y = c(0, Y1[x0:x1], 0),
  col = "#9ecae1",
  border = NA
)
polygon(
  x = c(X1[c(x1, x1:x2, x2)]),
  y = c(0, Y1[x1:x2], 0),
  col = "#deebf7",
  border = NA
)
polygon(
  x = c(X1[c(x2, x2:x3, x3)]),
  y = c(0, Y1[x2:x3], 0),
  col = "white",
  border = NA
)
polygon(
  x = c(X1[c(x3, x3:x4, x4)]),
  y = c(0, Y1[x3:x4], 0),
  col = "#deebf7",
  border = NA
)
polygon(
  x = c(X1[c(x4, x4:x5, x5)]),
  y = c(0, Y1[x4:x5], 0),
  col = "#9ecae1",
  border = NA
)
polygon(
  x = c(X1[c(x5, x5:200, 200)]),
  y = c(0, Y1[x5:200], 0),
  col = "#3182bd",
  border = NA
)

points(X1, Y1, type = "l")

text(0, Y1[x2] + 0.01, "66% of values", cex = 0.8)
lines(c(X1[x2], X1[x3]), c(Y1[x2] - 0.001, Y1[x3] - 0.001))

text(0, Y1[x1] + 0.01, "95% of values", cex = 0.8)
lines(c(X1[x1], X1[x4]), c(Y1[x1] - 0.001, Y1[x4] - 0.001))

text(0, Y1[x0] + 0.01, "99% of values", cex = 0.8)
lines(c(X1[x0], X1[x5]), c(Y1[x0] - 0.001, Y1[x5] - 0.001))

# arrows(2.816, 0.04, 2.816, dt(2.816, 18), length = 0.05)
#
# text(2.816, 0.06, "Location of\n our calculated\n value of t", cex = 0.7)


inst/tutorials/10 Linear models 2 Linear Regression/script_for_regression_videos.R
### Script for linear regression videos

# Load Chang et al data
grass <- read.csv(file.choose())

# Scatterplot plus regression line and residuals
par(mar = c(5,5,2,2))

plot(grass$BNPP_g_m2 ~ grass$Relative_biomass_grasses_percent, 
		 xlab = "Relative biomass of grasses %",
		 ylab = expression("Below ground primary productivity g.m" ^-2),
		 col = "aquamarine4",
		 pch = 16
)

mod1 <- lm(grass$BNPP_g_m2 ~ grass$Relative_biomass_grasses_percent)

# Regression line
abline(lm(grass$BNPP_g_m2 ~ grass$Relative_biomass_grasses_percent), 
			 lwd = 2, 
			 col = "aquamarine4")

# draw in residuals
arrows(x0 = grass$Relative_biomass_grasses_percent,
			 y0 = grass$BNPP_g_m2,
			 x1 = grass$Relative_biomass_grasses_percent,
			 y1 = mod1$fitted.values,
			 length = 0,
			 col = "aquamarine4"
)


# Scatterplot with poor fit line and residuals
plot(grass$BNPP_g_m2 ~ grass$Relative_biomass_grasses_percent, 
		 xlab = "Relative biomass of grasses %",
		 ylab = expression("Below ground primary productivity g.m" ^-2),
		 col = "aquamarine4",
		 pch = 16
)

abline(a = 180, 
			 b = -0.6,
			 lwd = 2,
			 col = "aquamarine4")

arrows(x0 = grass$Relative_biomass_grasses_percent,
			 y0 = grass$BNPP_g_m2,
			 x1 = grass$Relative_biomass_grasses_percent,
			 y1 = 180 - 0.6*grass$Relative_biomass_grasses_percent,
			 length = 0,
			 col = "aquamarine4"
)

# Plot with grand mean and residuals
plot(grass$BNPP_g_m2 ~ grass$rel_biomass_grass, 
		 xlab = "Relative biomass of grasses %",
		 ylab = expression("Below ground primary productivity g.m" ^-2),
		 col = "aquamarine4",
		 pch = 16
)

abline(h = mean(grass$BNPP_g_m2),
			 lwd = 2,
			 col = "aquamarine4")

arrows(x0 = grass$rel_biomass_grass,
			 y0 = grass$BNPP_g_m2,
			 x1 = grass$rel_biomass_grass,
			 y1 = mean(grass$BNPP_g_m2),
			 length = 0,
			 col = "aquamarine4"
)



# Changing some names so that the code is less messy
names(grass)[9] <- "biomass_grass"

productivity <- data.frame(grass$BNPP_g_m2, grass$biomass_grass)
names(productivity) <- c("BNPP_g_m2", "rel_biomass_grass")

# Calculate means etc.
mean(productivity$BNPP_g_m2)
sd(productivity$BNPP_g_m2)

mean(productivity$rel_biomass_grass)
sd(productivity$rel_biomass_grass)

cor(productivity$BNPP_g_m2, 
		productivity$rel_biomass_grass)

# Regression model 
mod1 <- lm(BNPP_g_m2 ~ rel_biomass_grass, 
					 data = productivity)

summary(mod1)$coefficients

# Scatterplot with means of x and y
plot(grass$BNPP_g_m2 ~ grass$biomass_grass, 
		 xlab = "Relative biomass of grasses %",
		 ylab = expression("Below ground primary productivity g.m" ^-2),
		 col = "aquamarine4",
		 pch = 16
)

abline(h = mean(grass$BNPP_g_m2),
			 lwd = 2,
			 col = "aquamarine4")

arrows(x0 = grass$biomass_grass,
			 y0 = grass$BNPP_g_m2,
			 x1 = grass$biomass_grass,
			 y1 = mean(grass$BNPP_g_m2),
			 length = 0,
			 col = "aquamarine4"
)


# Calculate sums of squares
SSTotal <- sum((productivity$BNPP_g_m2 - 
		mean(productivity$BNPP_g_m2))^2)
SSTotal

predicted <- 215 - 1.68 * productivity$rel_biomass_grass

SSError <- sum((productivity$BNPP_g_m2 - 
	  predicted)^2)						 
SSError


SSTotal - SSError


inst/tutorials/13 Linear models 1 Single factor ANOVA/ANOVA coefficients table video script.R
# ANOVA coefficients table video script

# Simple one-way

set.seed(210)

Treatment <- as.factor(rep(c("Control", "A", "B", "C"), each = 30))

Systolic_pressure <- c(rnorm(30, 150, sd = 4),
                       rnorm(30, 148, sd = 4),
                       rnorm(30, 145, sd = 5),
                       rnorm(30, 150, sd = 4))

boxplot(Systolic_pressure ~ Treatment,
        ylab = "Systolic blood pressure (mmHg)",
        col = "aquamarine4")

summary(lm(Systolic_pressure ~ Treatment))

Treatment <- relevel(Treatment, ref = "Control")

summary(lm(Systolic_pressure ~ Treatment))


# Two factor without interaction

set.seed(213)

Treatment <- as.factor(rep(c("Control", "A", "B", "C"), each = 40))

Age_class <- as.factor(rep(c("40-50", "70-80"), times = 80))

Systolic_pressure <- c(rnorm(40, 150, sd = 4),
                       rnorm(40, 149, sd = 4),
                       rnorm(40, 145, sd = 5),
                       rnorm(40, 153, sd = 4))

Systolic_pressure <- ifelse(Age_class == "70-80", Systolic_pressure + runif(1, min = 3, max = 5), Systolic_pressure)

boxplot(Systolic_pressure ~ Age_class*Treatment,
        ylab = "Systolic blood pressure (mmHg)",
        col = c("aquamarine4", "darkorange"))

summary(lm(Systolic_pressure ~ Treatment+Age_class))

Treatment <- relevel(Treatment, ref = "Control")

summary(lm(Systolic_pressure ~ Treatment + Age_class))

tapply(Systolic_pressure, 
       INDEX = list(Treatment, Age_class), 
       FUN = mean)



# Two factor with interaction

set.seed(20)

Treatment <- as.factor(rep(c("Control", "A", "B", "C"), each = 40))

Age_class <- as.factor(rep(c("40-50", "70-80"), times = 80))

Systolic_pressure <- c(rnorm(40, 150, sd = 4),
                       rnorm(40, 148, sd = 4),
                       rnorm(40, 145, sd = 5),
                       rnorm(40, 150, sd = 4))

Systolic_pressure <- ifelse(Age_class == "70-80" & Treatment != "B", Systolic_pressure + runif(1, min = 0, max = 10), Systolic_pressure)
Systolic_pressure <- ifelse(Age_class == "70-80" & Treatment == "B", Systolic_pressure + runif(1, min = 0, max = 2), Systolic_pressure)

boxplot(Systolic_pressure ~ Age_class*Treatment,
        ylab = "Systolic blood pressure (mmHg)",
        col = c("aquamarine4", "darkorange"))

summary(lm(Systolic_pressure ~ Treatment*Age_class))

Treatment <- relevel(Treatment, ref = "Control")

summary(lm(Systolic_pressure ~ Treatment * Age_class))


tapply(Systolic_pressure, 
       INDEX = list(Treatment, Age_class), 
       FUN = mean)


inst/tutorials/13 Linear models 1 Single factor ANOVA/Code for ANOVA video.R
library(ggplot2)
library(cowplot)
set.seed = 21

# Generate factor with three levels
F1 <- as.factor(rep(c("A", "B", "C"), each = 250))



# Generate data from three normal distributions
# sd = 3, means  = 19, 20, 20.5 depending on the level
# of F1

var1 <- c(rnorm(250, mean = 20, sd = 3),
          rnorm(250, mean = 20, sd = 3),
          rnorm(250, mean = 20, sd = 3))


# Generate data from three normal distributions
# sd = 3, means  = 15, 20, 23 depending on the level
# of F1

var2 <- c(rnorm(250, mean = 15, sd = 3),
          rnorm(250, mean = 20, sd = 3),
          rnorm(250, mean = 23, sd = 3))


p1 <- ggplot(data = data.frame(F1, var1),
             aes(x = var1, fill = F1)) +
  xlim(5, 35) +
  geom_density(alpha = 0.6) +
  theme_bw() +
  labs(y = "Density")




p2 <- ggplot(data = data.frame(F1, var2),
             aes(x = var2, fill = F1)) +
  xlim(5, 35) +
  geom_density(alpha = 0.6) +
  theme_bw() +
  labs(y = "Density")

plot_grid(p1, p2, nrow = 2)

var(var1)
var(var2)

SStotal1 <- sum((var2 - mean(var2))^2)

SSerror1 <- sum((var2[F1 == "A"] - mean(var2[F1 == "A"]))^2) +
            sum((var2[F1 == "B"] - mean(var2[F1 == "B"]))^2) +
            sum((var2[F1 == "C"] - mean(var2[F1 == "C"]))^2)

SStreatment1 <- SStotal1 - SSerror1


A1 <- lm(var1 ~ F1)
A2 <- lm(var2 ~ F1)

set.seed(2116)
F2 <- as.factor(rep(c("A", "B", "C"), each = 8))

var3 <- c(rnorm(8, mean = 20, sd = 2),
          rnorm(8, mean = 16, sd = 2),
          rnorm(8, mean = 22, sd = 2))

palette1 <- c("chocolate4", "aquamarine4", "firebrick3")

plot(var3, col = palette1[F2],
     pch= 16,
     xaxt = "n",
     xlab = "",
     ylab = "value")

legend(
  "bottomleft",
  legend = c("A", "B", "C"),
  pch = 16,
  col = palette1
)


abline(h = mean(var3), lwd = 2)

arrows(x0 = 1:24, 
       y0 = var3, 
       x1 = 1:24,
       y1 = rep(mean(var3),24),
       length = 0,
       col = palette1[F2])

plot(var3, col = palette1[F2],
     pch= 16,
     xaxt = "n",
     xlab = "",
     ylab = "Value")

legend(
  "bottomleft",
  legend = c("A", "B", "C"),
  pch = 16,
  col = palette1
)

mA <- mean(var3[F2 == "A"])
mB <- mean(var3[F2 == "B"])
mC <- mean(var3[F2 == "C"])

lines(x = c(1,8), y = c(mA, mA), lwd = 2, col = palette1[1])
lines(x = c(9,16), y = c(mB, mB), lwd = 2, col = palette1[2])
lines(x = c(17,24), y = c(mC, mC), lwd = 2, col = palette1[3])

arrows(x0 = 1:24, 
       y0 = var3, 
       x1 = 1:24,
       y1 = rep(c(mA, mB, mC), each = 8),
       length = 0,
       col = palette1[F2])     
